app/models/commercial.rb

```diff
@@ -1,10 +1,43 @@
 class Commercial < ActiveRecord::Base
+  require 'oembed'
+
   belongs_to :commercialable, polymorphic: true
 
-  attr_accessible :commercial_id, :commercial_type
+  attr_accessible :url
+
+  validates :url, presence: true
+  validates :url, format: URI::regexp(%w(http https))
+
+  def self.provider
+    {
+      youtube: 'YouTube',
+      slideshare: 'SlideShare',
+      flickr: 'Flickr',
+      vimeo: 'Vimeo',
+      speakerdeck: 'Speakerdeck',
+      instagram: 'Instagram'
+    }
+  end
+
+  def self.get_content(url)
+    speakerdeck = OEmbed::Provider.new('http://speakerdeck.com/oembed.json')
+    speakerdeck << 'https://speakerdeck.com/*'
+    speakerdeck << 'http://speakerdeck.com/*'
 
-  validates :commercial_id, presence: true
-  validates :commercial_type, presence: true
+    OEmbed::Providers.register(
+        OEmbed::Providers::Youtube,
+        OEmbed::Providers::Vimeo,
+        OEmbed::Providers::Slideshare,
+        OEmbed::Providers::Flickr,
+        OEmbed::Providers::Instagram,
+        speakerdeck
+    )
 
-  validates :commercial_type, inclusion: { in: CONFIG['commercial_types'].values }
+    begin
+      resource = OEmbed::Providers.get(url, maxwidth: 560, maxheight: 315)
+      resource.html
+    rescue StandardError
+      ''
```

Comment:
OEmbed::Providers.get throws an Error when you enter an invalid url (not supported provider eg) and then I return an empty string to display nothing in the view! 
But yes, we can discuss about a better approach :) Any suggestion?